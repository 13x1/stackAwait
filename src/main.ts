import KVIN from '@/kvin2';

type AU = Array<unknown>;

type PromiseResult<T> = {
    state: "fulfilled",
    value: T,
    promise: Promise<T>
} | {
    state: "rejected",
    reason: Error,
    promise: Promise<T>
} | {
    state: "pending",
    promise: Promise<T>
}

interface FunctionalCallData {
    fn: (...args: AU) => unknown;
    args: AU;
    vArgs: unknown;
    bThis: unknown;
}

export type PromiseMap = Map<string, PromiseResult<unknown>>

interface StackAwaitOpts<This = undefined> {
    /**
     * Serialize an async function call to a string.
     */
    serializer: (data: FunctionalCallData) => string;
    /**
     * The virtual arguments to the function.
     */
    vArgs: unknown;
    /**
     * The async scope object to use.
     * Pass `false` or `undefined` to disable.
     */
    asyncScope: undefined | false | PromiseMap
    /**
     * Bind (`this`) argument
     */
    bThis: This,
    /**
     * Whether to throw or return the rejection as a value with a
     * marker to be used with {@link isRejection}. Default is `value`,
     * because try/catch is annoying to use given that we abuse it ourselves.
     */
    rejections: "value" | "throw"
}

export const stackAwaitOptsDefaults: StackAwaitOpts = {
    serializer: data => KVIN.stringify(data),
    vArgs: undefined,
    asyncScope: undefined,
    bThis: undefined,
    rejections: "value"
};

let nullCache = new Map();
let currentGlobalCache: PromiseMap = nullCache

function logAndThrow(m: unknown) {
    if (!__debug__.disable_logging) console.error(m)
    throw m
}

export function stackAwait<Args extends AU, Return, This = undefined>(
    opts: Partial<StackAwaitOpts<This>>,
    fn: (this: This, ...args: Args) => Promise<Return>,
    ...args: Args
): Return;
export function stackAwait<Args extends AU, Return>(
    fn: (...args: Args) => Promise<Return>,
    ...args: Args
): Return;
export function stackAwait(...args: unknown[]) {
    const o = typeof args[0] === 'object' ? args.shift() : {};
    const fn = args.shift() as (...args: unknown[]) => Promise<unknown>;
    if (__debug__.return_args) return {opts: o, func: fn, args};
    const options = Object.assign({}, stackAwaitOptsDefaults, o);
    if (__debug__.return_proc_args) return options;

    const cache: PromiseMap = options.asyncScope || currentGlobalCache
    if (cache === nullCache) logAndThrow(new Error("Illegal invocation: Not called inside runAsyncStack() or custom scope"))

    let serialized = options.serializer({
        fn, args, vArgs: options.vArgs, bThis: options.bThis
    })

    // magic happens here
    let cacheHit = cache.get(serialized)
    if (!cacheHit) {
        cache.set(serialized, {
            state: 'pending',
            promise: fn.apply(options.bThis, args)
        })
        throw stackAwaitError
    } else if (cacheHit.state === 'pending') {
        // usually this shouldn't happen, but still
        throw stackAwaitError
    } else if (cacheHit.state === 'rejected') {
        if (options.rejections === "throw") throw genRejection(cacheHit.reason)
        return genRejection(cacheHit.reason)
    } else if (cacheHit.state === "fulfilled") {
        return cacheHit.value
    }
}

interface RunAsyncStackOpts<This = undefined> {
    /**
     * The maximum amount iterations to run before giving up.
     * @see Halting problem
     */
    maxIterations: number;
    /**
     * The async scope object to use.
     * Pass `false` or `undefined` to disable.
     */
    asyncScope: undefined | false | PromiseMap,
    /**
     * Bind (`this`) argument
     */
    bThis: This,
}

export const runAsyncStackOptsDefaults: RunAsyncStackOpts = {
    maxIterations: 500,
    asyncScope: undefined,
    bThis: undefined,
}

export async function runAsyncStack<Args extends AU, Return, This = undefined>(
    opts: Partial<RunAsyncStackOpts<This>>,
    fn: (this: This, ...args: Args) => Return,
    ...args: Args
): Promise<Return>;
export async function runAsyncStack<Args extends AU, Return>(
    fn: (...args: Args) => Return,
    ...args: Args
): Promise<Return>;
export async function runAsyncStack(...args: unknown[]) {
    const o = typeof args[0] === 'object' ? args.shift() : {};
    const fn = args.shift() as (...args: unknown[]) => unknown;
    const options = Object.assign({}, runAsyncStackOptsDefaults, o);
    const localCache = options.asyncScope || new Map()
    for (let att = 0; att < options.maxIterations; att++) {
        let returnValue = null
        let error = null
        let caught = false
        const prevCache = currentGlobalCache
        currentGlobalCache = localCache
        // collect promise
        try {
            returnValue = fn.apply(options.bThis, args)
        } catch (e) {
            caught = true
            error = e
        }
        currentGlobalCache = prevCache
        if (returnValue instanceof Promise) {
            try {
                returnValue = await returnValue
            } catch (e) {
                caught = true
                error = e
            }
        }
        if (caught && error !== stackAwaitError) {
            throw error
        }
        if (caught) {
            // resolve cache
            for (let [_, entry] of localCache) {
                if (entry.state === "pending") {
                    try {
                        entry.state = "fulfilled"
                        entry.value = await entry.promise
                    } catch (e) {
                        entry.state = "rejected"
                        entry.reason = e
                    }
                }
            }
            continue
        }
        return returnValue
    }
    throw new Error("Maximum iterations reached!")
}

let rejectionSymbol = Symbol("Promise rejection")

export function isRejection(r: unknown): r is Error {
    return (r as any)[rejectionSymbol] !== undefined
}

function genRejection(e: Error) {
    (e as any)[rejectionSymbol] = true
    if (!isRejection(e)) return {value: e, [rejectionSymbol]: true}
    return e
}

let errDesc = "StackAwait error. If you caught this, please see the docs.";
let stackAwaitErrorSymbol = Symbol(errDesc)

let stackAwaitError = {
    [stackAwaitErrorSymbol]: true,
    toString: () => errDesc
}

export function enforceStackThrows(e: unknown) {
    // let err = (e as any)[stackAwaitErrorSymbol]
    if (e === stackAwaitError) throw e
}

export const __debug__ = {
    return_args: false,
    return_proc_args: false,
    disable_logging: false,
    stackAwaitError, errDesc
}