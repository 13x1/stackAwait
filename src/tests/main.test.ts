import KVIN from '@/kvin2';
import { describe, it, expect, assert } from 'vitest';
import { stackAwait, stackAwaitOptsDefaults } from '@/main.js';

describe('Types', () => {
    it('should compile', () => null);
    if (Math.random() > 0) return;

    assert(stackAwait(fetch, 'test') satisfies Response);
    assert(stackAwait({}, fetch, 'test', {method: 'POST'}) satisfies Response);
});

describe('Serialization', () => {
    const ser = (arg: unknown) => stackAwaitOptsDefaults.serializer(() => null, [], arg);

    function deser(str: string): unknown {
        return (KVIN.deserialize(str) as { vArgs: unknown }).vArgs;
    }

    function test(gen: () => Array<unknown>, snap = true, mapper = (e: unknown) => e) {
        expect(ser(gen()), 'serialization').toBe(ser(gen()));
        expect(gen().map(mapper), 'equality').toEqual(gen().map(mapper));
        expect(gen().map(mapper), 're-serialized equality').toStrictEqual(
            (deser(ser(gen())) as Array<unknown>).map(mapper)
        );
        if (snap) expect(ser(gen())).toMatchSnapshot();
    }

    it('should serialize JSON primitives', () => {
        test(() => ['test', 1, true, {test: 'test'}, ['test', 1, true, {test: 'test'}]]);
    });

    it('should serialize undefined, null, NaN, Infinity, -Infinity', () => {
        test(() => [undefined, null, NaN, Infinity, -Infinity]);
    });

    it('should serialize Typed Arrays (Float64Array, Int64Array, etc.) and Maps & Sets', () => {
        test(() => [
            new Float64Array([1, 2, 3]),
            new Uint16Array([1, 2, 3]),
            new Map([['test', 'test']]),
            new Set(['test', '123'])
        ]);
    });

    it('should serialize Object graphs with cycles, Arrays with enumerable, non-numeric properties and sparse arrays', () => {
        test(() => {
            // Object graphs with cycles
            const obj = {test: 'test'};
            // eslint-disable-next-line
            (obj as any).obj = obj;
            // Arrays with enumerable, non-numeric properties and sparse arrays
            const arr = ['test', 1, true, {test: 'test'}];
            // eslint-disable-next-line
            (arr as any).test = 'test';
            arr[10] = 'test';
            return [obj, arr];
        });
    });

    it('should serialize Date, URL, Error, BigInt, RegExp and boxed primitives like String and Number', () => {
        const date = new Date('2023-07-23T19:33:45.093Z');
        // noinspection JSPrimitiveTypeWrapperUsage
        test(() => [
            date,
            new URL('https://example.com'),
            BigInt(123),
            /test/,
            new String('test'),
            new Number(123)
        ]);
        // errors tend to be a problem child bc stack traces can be different
        const err = new Error('test');
        test(() => [err], false);
    });

    it('should serialize functions', () => {
        test(
            () => {
                async function* test1() {
                    await new Promise(resolve => setTimeout(resolve, 1000));
                    yield 'test';
                }

                return [
                    test1,
                    async function* test2() {
                        await new Promise(resolve => setTimeout(resolve, 1000));
                        yield 'test';
                    },
                    async function* () {
                        yield 'test';
                    },
                    async () => 5,
                    () => {
                        return 5;
                    }
                ];
            },
            true,
            e => e.toString()
        );
    });

    it('should fail to serialize closures', () => {
        test(
            () => {
                const n = Math.random();

                async function* test1() {
                    await new Promise(resolve => setTimeout(resolve, 1000));
                    yield n;
                }

                return [
                    test1,
                    async function* test2() {
                        await new Promise(resolve => setTimeout(resolve, 1000));
                        yield n;
                    },
                    async function* () {
                        yield n;
                    },
                    async () => n,
                    () => {
                        return n;
                    }
                ];
            },
            true,
            e => e.toString()
        );
    });

    it('should fail to serialize promises', () => {
        test(
            () => {
                const n = Math.random();
                const p = new Promise(resolve => resolve(n));
                return [p];
            },
            true,
            e => e.toString()
        );
    });

    it('should fail to serialize objects like Response', () => {
        test(
            () => {
                return [new Response(Math.random().toString())];
            },
            true,
            e => e.toString()
        );
    });
});
