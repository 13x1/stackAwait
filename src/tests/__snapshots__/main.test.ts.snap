// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Serialization > should fail to serialize closures 1`] = `
"{
    \\"_serializeVerId\\": \\"v8\\",
    \\"what\\": {
        \\"ctr\\": 0,
        \\"ps\\": {
            \\"fn\\": {
                \\"ctr\\": 15,
                \\"ps\\": {},
                \\"fnName\\": \\"\\",
                \\"arg\\": \\"() => null\\"
            },
            \\"args\\": {
                \\"raw\\": []
            },
            \\"vArgs\\": {
                \\"arr\\": [
                    {
                        \\"ctr\\": \\"AsyncGeneratorFunction\\",
                        \\"ps\\": {},
                        \\"fnName\\": \\"test1\\",
                        \\"arg\\": \\"async function* test1() {\\\\n          await new Promise((resolve) => setTimeout(resolve, 1e3));\\\\n          yield n;\\\\n        }\\"
                    },
                    {
                        \\"ctr\\": \\"AsyncGeneratorFunction\\",
                        \\"ps\\": {},
                        \\"fnName\\": \\"test2\\",
                        \\"arg\\": \\"async function* test2() {\\\\n            await new Promise((resolve) => setTimeout(resolve, 1e3));\\\\n            yield n;\\\\n          }\\"
                    },
                    {
                        \\"ctr\\": \\"AsyncGeneratorFunction\\",
                        \\"ps\\": {},
                        \\"fnName\\": \\"\\",
                        \\"arg\\": \\"async function* () {\\\\n            yield n;\\\\n          }\\"
                    },
                    {
                        \\"ctr\\": \\"AsyncFunction\\",
                        \\"ps\\": {},
                        \\"fnName\\": \\"\\",
                        \\"arg\\": \\"async () => n\\"
                    },
                    {
                        \\"ctr\\": 15,
                        \\"ps\\": {},
                        \\"fnName\\": \\"\\",
                        \\"arg\\": \\"() => {\\\\n            return n;\\\\n          }\\"
                    }
                ]
            }
        }
    }
}"
`;

exports[`Serialization > should fail to serialize objects like Response 1`] = `
"{
    \\"_serializeVerId\\": \\"v8\\",
    \\"what\\": {
        \\"ctr\\": 0,
        \\"ps\\": {
            \\"fn\\": {
                \\"ctr\\": 15,
                \\"ps\\": {},
                \\"fnName\\": \\"\\",
                \\"arg\\": \\"() => null\\"
            },
            \\"args\\": {
                \\"raw\\": []
            },
            \\"vArgs\\": {
                \\"arr\\": [
                    {
                        \\"ctr\\": \\"Response\\",
                        \\"ps\\": {}
                    }
                ]
            }
        }
    }
}"
`;

exports[`Serialization > should fail to serialize promises 1`] = `
"{
    \\"_serializeVerId\\": \\"v8\\",
    \\"what\\": {
        \\"ctr\\": 0,
        \\"ps\\": {
            \\"fn\\": {
                \\"ctr\\": 15,
                \\"ps\\": {},
                \\"fnName\\": \\"\\",
                \\"arg\\": \\"() => null\\"
            },
            \\"args\\": {
                \\"raw\\": []
            },
            \\"vArgs\\": {
                \\"arr\\": [
                    {
                        \\"resolve\\": {
                            \\"_serializeVerId\\": \\"v8\\",
                            \\"what\\": \\"<Unresolved Promise>\\"
                        }
                    }
                ]
            }
        }
    }
}"
`;

exports[`Serialization > should serialize Date, URL, Error, BigInt, RegExp and boxed primitives like String and Number 1`] = `
"{
    \\"_serializeVerId\\": \\"v8\\",
    \\"what\\": {
        \\"ctr\\": 0,
        \\"ps\\": {
            \\"fn\\": {
                \\"ctr\\": 15,
                \\"ps\\": {},
                \\"fnName\\": \\"\\",
                \\"arg\\": \\"() => null\\"
            },
            \\"args\\": {
                \\"raw\\": []
            },
            \\"vArgs\\": {
                \\"arr\\": [
                    {
                        \\"ctr\\": 19,
                        \\"ps\\": {},
                        \\"arg\\": \\"2023-07-23T19:33:45.093Z\\"
                    },
                    {
                        \\"ctr\\": 18,
                        \\"ps\\": {},
                        \\"arg\\": \\"https://example.com/\\"
                    },
                    {
                        \\"bigint\\": \\"123\\"
                    },
                    {
                        \\"ctr\\": 10,
                        \\"arg\\": \\"test\\"
                    },
                    {
                        \\"ctr\\": 12,
                        \\"arg\\": \\"test\\"
                    },
                    {
                        \\"ctr\\": 11,
                        \\"arg\\": \\"123\\"
                    }
                ]
            }
        }
    }
}"
`;

exports[`Serialization > should serialize JSON primitives 1`] = `
"{
    \\"_serializeVerId\\": \\"v8\\",
    \\"what\\": {
        \\"ctr\\": 0,
        \\"ps\\": {
            \\"fn\\": {
                \\"ctr\\": 15,
                \\"ps\\": {},
                \\"fnName\\": \\"\\",
                \\"arg\\": \\"() => null\\"
            },
            \\"args\\": {
                \\"raw\\": []
            },
            \\"vArgs\\": {
                \\"arr\\": [
                    \\"test\\",
                    1,
                    true,
                    {
                        \\"raw\\": {
                            \\"test\\": \\"test\\"
                        }
                    },
                    {
                        \\"arr\\": [
                            \\"test\\",
                            1,
                            true,
                            {
                                \\"raw\\": {
                                    \\"test\\": \\"test\\"
                                }
                            }
                        ]
                    }
                ]
            }
        }
    }
}"
`;

exports[`Serialization > should serialize Object graphs with cycles, Arrays with enumerable, non-numeric properties and sparse arrays 1`] = `
"{
    \\"_serializeVerId\\": \\"v8\\",
    \\"what\\": {
        \\"ctr\\": 0,
        \\"ps\\": {
            \\"fn\\": {
                \\"ctr\\": 15,
                \\"ps\\": {},
                \\"fnName\\": \\"\\",
                \\"arg\\": \\"() => null\\"
            },
            \\"args\\": {
                \\"raw\\": []
            },
            \\"vArgs\\": {
                \\"arr\\": [
                    {
                        \\"ctr\\": 0,
                        \\"ps\\": {
                            \\"test\\": {
                                \\"raw\\": \\"test\\"
                            },
                            \\"obj\\": {
                                \\"seen\\": 3
                            }
                        }
                    },
                    {
                        \\"arr\\": [
                            \\"test\\",
                            1,
                            true,
                            {
                                \\"raw\\": {
                                    \\"test\\": \\"test\\"
                                }
                            }
                        ],
                        \\"isl\\": [
                            {
                                \\"@\\": 0,
                                \\"arr\\": [
                                    \\"test\\",
                                    1,
                                    true,
                                    {
                                        \\"test\\": \\"test\\"
                                    }
                                ]
                            }
                        ],
                        \\"ps\\": {
                            \\"10\\": \\"test\\",
                            \\"test\\": \\"test\\"
                        },
                        \\"len\\": 11
                    }
                ]
            }
        }
    }
}"
`;

exports[`Serialization > should serialize Typed Arrays (Float64Array, Int64Array, etc.) and Maps & Sets 1`] = `
"{
    \\"_serializeVerId\\": \\"v8\\",
    \\"what\\": {
        \\"ctr\\": 0,
        \\"ps\\": {
            \\"fn\\": {
                \\"ctr\\": 15,
                \\"ps\\": {},
                \\"fnName\\": \\"\\",
                \\"arg\\": \\"() => null\\"
            },
            \\"args\\": {
                \\"raw\\": []
            },
            \\"vArgs\\": {
                \\"arr\\": [
                    {
                        \\"ctr\\": 9,
                        \\"ab8\\": \\"\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000รฐ?\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000@\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\b@\\"
                    },
                    {
                        \\"ctr\\": 5,
                        \\"ab8\\": \\"\\\\u0001\\\\u0000\\\\u0002\\\\u0000\\\\u0003\\\\u0000\\"
                    },
                    {
                        \\"mapKeys\\": {
                            \\"arr\\": [
                                \\"test\\"
                            ]
                        },
                        \\"mapVals\\": {
                            \\"arr\\": [
                                \\"test\\"
                            ]
                        }
                    },
                    {
                        \\"setVals\\": {
                            \\"arr\\": [
                                \\"test\\",
                                \\"123\\"
                            ]
                        }
                    }
                ]
            }
        }
    }
}"
`;

exports[`Serialization > should serialize functions 1`] = `
"{
    \\"_serializeVerId\\": \\"v8\\",
    \\"what\\": {
        \\"ctr\\": 0,
        \\"ps\\": {
            \\"fn\\": {
                \\"ctr\\": 15,
                \\"ps\\": {},
                \\"fnName\\": \\"\\",
                \\"arg\\": \\"() => null\\"
            },
            \\"args\\": {
                \\"raw\\": []
            },
            \\"vArgs\\": {
                \\"arr\\": [
                    {
                        \\"ctr\\": \\"AsyncGeneratorFunction\\",
                        \\"ps\\": {},
                        \\"fnName\\": \\"test1\\",
                        \\"arg\\": \\"async function* test1() {\\\\n          await new Promise((resolve) => setTimeout(resolve, 1e3));\\\\n          yield \\\\\\"test\\\\\\";\\\\n        }\\"
                    },
                    {
                        \\"ctr\\": \\"AsyncGeneratorFunction\\",
                        \\"ps\\": {},
                        \\"fnName\\": \\"test2\\",
                        \\"arg\\": \\"async function* test2() {\\\\n            await new Promise((resolve) => setTimeout(resolve, 1e3));\\\\n            yield \\\\\\"test\\\\\\";\\\\n          }\\"
                    },
                    {
                        \\"ctr\\": \\"AsyncGeneratorFunction\\",
                        \\"ps\\": {},
                        \\"fnName\\": \\"\\",
                        \\"arg\\": \\"async function* () {\\\\n            yield \\\\\\"test\\\\\\";\\\\n          }\\"
                    },
                    {
                        \\"ctr\\": \\"AsyncFunction\\",
                        \\"ps\\": {},
                        \\"fnName\\": \\"\\",
                        \\"arg\\": \\"async () => 5\\"
                    },
                    {
                        \\"ctr\\": 15,
                        \\"ps\\": {},
                        \\"fnName\\": \\"\\",
                        \\"arg\\": \\"() => {\\\\n            return 5;\\\\n          }\\"
                    }
                ]
            }
        }
    }
}"
`;

exports[`Serialization > should serialize undefined, null, NaN, Infinity, -Infinity 1`] = `
"{
    \\"_serializeVerId\\": \\"v8\\",
    \\"what\\": {
        \\"ctr\\": 0,
        \\"ps\\": {
            \\"fn\\": {
                \\"ctr\\": 15,
                \\"ps\\": {},
                \\"fnName\\": \\"\\",
                \\"arg\\": \\"() => null\\"
            },
            \\"args\\": {
                \\"raw\\": []
            },
            \\"vArgs\\": {
                \\"arr\\": [
                    {
                        \\"undefined\\": true
                    },
                    {
                        \\"raw\\": null
                    },
                    {
                        \\"number\\": \\"NaN\\"
                    },
                    {
                        \\"number\\": \\"Infinity\\"
                    },
                    {
                        \\"number\\": \\"-Infinity\\"
                    }
                ]
            }
        }
    }
}"
`;
